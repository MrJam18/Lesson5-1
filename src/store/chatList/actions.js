import { onValue, push, remove, set } from "firebase/database";
import { chatListRef, getMsgsRefById } from "../../service/firebase";

export const addChat = (name, chatID) => ({
    type: 'CHAT_LIST::ADD_CHAT',
    chatID, name
})
export const deleteChat = (chatID) => ({
    type: "CHAT_LIST::DELETE_CHAT",
    chatID
})

export const changeChatListAction = (chatList) => ({
    type: "CHAT_LIST::CHANGE",
    chatList
})
    let newChatID = 1;
export const addChatThunk = (chatName) => () => {
    newChatID++;
    const fullChatID = 'chat' + newChatID;
    push(chatListRef, {name: chatName, id: fullChatID, img: 'default.png'});
    set(getMsgsRefById(fullChatID), {empty: true});
}
export const deleteChatThunk = (chatID) => (_, getState) => {
    let newChatList = [...getState().chats];
    newChatList = newChatList.filter(elem => elem.id !== chatID);
    set(chatListRef, newChatList);
    remove(getMsgsRefById(chatID));  
}
export const iniChatListTrackingThunk = () => (dispatch) => {
    onValue(chatListRef, (snapshot) => {
        let chatListArray = [];
        snapshot.forEach((el) => {
         chatListArray.push(el.val());
      });
      dispatch(changeChatListAction(chatListArray));
    });
    
}