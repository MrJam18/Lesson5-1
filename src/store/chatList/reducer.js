const initialChatList = [
    {name: 'John Snow', id: 'chat1', img: 'chat1.jpg'},
    {name: "Zhanna D'ark", id: 'chat2', img: 'chat2.jpg'},
    {name: 'Geralt from Rivia', id: 'chat3', img: 'chat3.jpg'},
    {name: "Eniken Skywalker", id: 'chat4', img: 'chat4.jpg'},
    {name: "bot", id: 'chat5', img: 'chat5.jpg'},
]
export const chatListReducer = (state = initialChatList, action) => {
    switch(action.type) {
        case 'CHAT_LIST::ADD_CHAT':
            return [
                ...state,
                {name: action.name, id: action.chatID, img: 'default.png'}
            ]
        case "CHAT_LIST::DELETE_CHAT":
            return state.filter((elem) => elem.id !== action.chatID);
        case 'CHAT_LIST::CHANGE':
            return action.chatList;
        default: return state;
    }
}