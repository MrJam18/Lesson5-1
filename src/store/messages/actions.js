export const addMessage = (chatID, message, author) => ({
    type: 'MESSAGES::ADD_MESSAGE',
    chatID, message, author
})
export const addMessageList = (chatID) => ({
    type: 'MESSAGES::ADD_MESSAGE_LIST',
    chatID
})