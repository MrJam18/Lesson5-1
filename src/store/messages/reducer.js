
const initialMessageList = {
    chat1: [],
    chat2: [],
    chat3: [],
    chat4: [],
    chat5: [],
}
export const messagesReducer = (state = initialMessageList, action) => {
    switch(action.type) {
        case 'MESSAGES::ADD_MESSAGE': 
        return ({
            ...state,
             [action.chatID]: [...state[action.chatID], {text: action.message, id: action.id, author: action.author}]
        })
        case 'MESSAGES::ADD_MESSAGE_LIST':
        return ({
            ...state,
            [action.chatID]: []
        })
        case 'MESSAGES::DELETE_MESSAGE_LIST':
            delete state[action.chatID]
            return ({
                ...state
            })
        default: return state;
    }
}