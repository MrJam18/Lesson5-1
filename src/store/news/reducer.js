const initState = {
    news: [],
    error: false,
    loading: true
}
export const newsReducer = (state = initState, action) => {
    switch(action.type) {
        case 'NEWS::CHANGE_NEWS': 
        return ({ ...state,
                news: action.news
         })
        case 'NEWS::SET_ERROR':
            return ({
                ...state, 
                error: action.error
            })
        case 'NEWS::SET_LOADING': 
            return ({
                ...state,
                loading: action.loading
            })
         default: return state;
    }
}