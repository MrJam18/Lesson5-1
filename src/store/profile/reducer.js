const initialProfile = {
    showData: true,
    name: 'User',
    surname: 'Default',
    birthDate: '1994-04-17',
    gender: "Мужской",
    isAuthed: false
    }

export const profileReducer = (state = initialProfile, action) => {
    switch (action.type) {
        case 'PROFILE::CHANGE_DATA': 
        return {
            ...state, 
            name: action.payload.name,
            surname: action.payload.surname,
            birthDate: action.payload.birthDate,
            gender: action.payload.gender,
            showData: action.payload.showData
        }
        case 'PROFILE::LOG_OUT':
            return {
                ...state,
                isAuthed: false
            }
        case 'PROFILE::LOG_IN':
            return {
                ...state,
                isAuthed: true
            }
        // case 'PROFILE::CREATE_USER':
        //     return {
        //         ...state,
        //         [action.payload.userID]: {
        //         name: action.payload.name,
        //         surname: action.payload.surname,
        //         birthDate: action.payload.birthDate,
        //         gender: action.payload.gender,
        //         showData: action.payload.showData
        //         },
        //         currentUser: action.payload.userID
        //     }
        default: return state;
    }
}