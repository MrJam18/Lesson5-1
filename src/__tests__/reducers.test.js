import { chatListReducer } from "../store/chatList/reducer"
import { messagesReducer } from "../store/messages/reducer"

describe('reducers tests', ()=> {
    it('return initState from message Reducer if type did not match', () => {
        const expected = {
            chat1: [],
            chat2: [],
            chat3: [],
            chat4: [],
            chat5: [],
        }
        const action = {
            type: 'MESSAGES::SOMETHING OTHER',
            payload: 'something'
        }
        const recieved = messagesReducer(expected, action);
        expect(recieved).toEqual(expected);
    })
    it("don't delete nothing if chat id did not match to state", ()=> {
        const state = [
            {name: 'John Snow', id: 'chat1', img: 'chat1.jpg'},
            {name: "Zhanna D'ark", id: 'chat2', img: 'chat2.jpg'},
            {name: 'Geralt from Rivia', id: 'chat3', img: 'chat3.jpg'},
            {name: "Eniken Skywalker", id: 'chat4', img: 'chat4.jpg'},
            {name: "bot", id: 'chat5', img: 'chat5.jpg'},
        ]
        const action = {
                type: "CHAT_LIST::DELETE_CHAT",
                chatID: 'chat10'
            }
        const expected = [
            {name: 'John Snow', id: 'chat1', img: 'chat1.jpg'},
            {name: "Zhanna D'ark", id: 'chat2', img: 'chat2.jpg'},
            {name: 'Geralt from Rivia', id: 'chat3', img: 'chat3.jpg'},
            {name: "Eniken Skywalker", id: 'chat4', img: 'chat4.jpg'},
            {name: "bot", id: 'chat5', img: 'chat5.jpg'},
        ]
        const recieved = chatListReducer(state, action);
        expect(recieved).toEqual(expected);
    })
    it('test add chatList from chatList Reducer with snapshots', ()=> {
        const state = [
            {name: 'John Snow', id: 'chat1', img: 'chat1.jpg'},
            {name: "Zhanna D'ark", id: 'chat2', img: 'chat2.jpg'},
            {name: 'Geralt from Rivia', id: 'chat3', img: 'chat3.jpg'},
            {name: "Eniken Skywalker", id: 'chat4', img: 'chat4.jpg'},
            {name: "bot", id: 'chat5', img: 'chat5.jpg'},
        ]
        const action = {
            type: 'CHAT_LIST::ADD_CHAT',
            chatID: 'chat6', name: 'Adriano Chelentano'
        }
        const recieved = chatListReducer(state, action);
        expect(recieved).toMatchSnapshot();
    })
})