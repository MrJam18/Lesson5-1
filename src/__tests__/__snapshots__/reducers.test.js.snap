// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers tests test add chatList from chatList Reducer with snapshots 1`] = `
Array [
  Object {
    "id": "chat1",
    "img": "chat1.jpg",
    "name": "John Snow",
  },
  Object {
    "id": "chat2",
    "img": "chat2.jpg",
    "name": "Zhanna D'ark",
  },
  Object {
    "id": "chat3",
    "img": "chat3.jpg",
    "name": "Geralt from Rivia",
  },
  Object {
    "id": "chat4",
    "img": "chat4.jpg",
    "name": "Eniken Skywalker",
  },
  Object {
    "id": "chat5",
    "img": "chat5.jpg",
    "name": "bot",
  },
  Object {
    "id": "chat6",
    "img": "default.png",
    "name": "Adriano Chelentano",
  },
]
`;
